keps/sig-node/5304-dra-attributes-downward-api/README.md

```diff
@@ -0,0 +1,883 @@
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes a Downward API for Dynamic Resource Allocation (DRA) device attributes, implemented entirely in the kubelet. Workloads like KubeVirt need device identifiers (e.g., PCIe bus address for physical GPUs, mediated device UUID for virtual GPUs) to configure device access inside guests. While these identifiers exist in DRA objects (`ResourceClaim`, `ResourceSlice`), they are not currently consumable via the Pod's Downward API. This proposal adds a new Downward API selector (`draDeviceFieldRef`) for environment variables and projected volumes. The kubelet will run a local DRA attributes controller that watches `ResourceClaim` and `ResourceSlice` objects, caches resolved attributes per Pod resource claim/request, and resolves `draDeviceFieldRef` on demand.
+
+## Motivation
+
+Workloads that need to interact with DRA-allocated devices (like KubeVirt virtual machines) require access to device-specific identifiers such as PCIe bus addresses or mediated device UUIDs. These attributes are already stored in DRA `ResourceClaim` and `ResourceSlice` objects, but there is currently no standardized way for Pods to consume them. Users must resort to custom controllers that watch these objects and inject attributes via annotations or custom resources, which is fragile and non-portable.
+
+The Kubernetes Downward API provides a standard mechanism for exposing Pod and container metadata to workloads. Extending this API to support DRA device attributes would enable workloads to discover device information without requiring additional custom controllers or privileged access to the Kubernetes API.
+
+### Goals
+
+- Provide a stable Downward API path for device attributes associated with `pod.spec.resourceClaims[*]` requests
+- Support any attributes defined in ResourceSlice objects, controlled by the allowlist mechanism
+- Allow resource owners to explicitly allowlist which attributes are exposed (maximum 8 per request) for security and control
+- Maintain compatibility with existing DRA drivers without requiring changes to driver interfaces
+
+### Non-Goals
+
+- Expose the entirety of `ResourceClaim`/`ResourceSlice` objects in the Downward API
+- Allow arbitrary JSONPath into external objects via Downward API
+- Change or extend DRA driver interfaces
+- Support dynamic updates to device attributes after Pod container startup (for env vars)
```

Comment:
Note that resource slices could change or disappear after allocation. We don't plan to handle that either, I suppose. Well, more accurately, we will take whatever is in the resource slice at the time of Pod creation, NOT at the time of allocation, IIUC. Otherwise you need scheduler side integration to copy the attributes from slice to claim at allocation time.